## Model predictions using the preprocessed data

In [4]:
import pandas as pd
from sklearn.svm import SVC
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score
from sklearn.neural_network import MLPClassifier

np.random.seed(1)

In [5]:
X_train = pd.read_csv("C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/Week7_Weekend_submission/X_train.csv")
X_test = pd.read_csv("C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/Week7_Weekend_submission/X_test.csv")
y_train = pd.read_csv("C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/Week7_Weekend_submission/y_train.csv")
y_test = pd.read_csv("C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/Week7_Weekend_submission/y_test.csv")

In [11]:
X_train.shape

(548, 1472)

In [12]:
X_test.shape

(58304, 1472)

In [28]:
y_train.shape

(548, 1)

In [29]:
y_test.shape

(58304, 1)

Moving forward I will be choosing F1 score to evaluate the performance of the model. I am choosing this scoring metric because it combines both precision and recall value into one scoring metric by defining the harmonic mean and works best for this kind of probelms.

# Initial random search based on the parameters set for the Random search for the logistic regression

In [5]:
score_measure = "f1"
kfolds = 3

param_grid = [
    { 'penalty': ['l2'], 'C': [ 0.1, 1, 10]}
]

Lr = LogisticRegression()
rand_search = RandomizedSearchCV(estimator = Lr, param_distributions=param_grid, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 3 is smaller than n_iter=100. Running 3 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 3 folds for each of 3 candidates, totalling 9 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9779602625902779
... with parameters: {'penalty': 'l2', 'C': 10}


## Grid search for the logistic regression

In [6]:
score_measure = "f1"
kfolds = 5

param_grid = [
    { 'penalty': ['l1', 'l2'], 'C': [0.01, 0.1, 1, 10, 100]}
]

Lr_grid = LogisticRegression()
grid_search = GridSearchCV(estimator = Lr_grid, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 10 candidates, totalling 50 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1461, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 447, in _check_solver
    raise Valu

The best f1 score is 0.9742068308729206
... with parameters: {'C': 10, 'penalty': 'l2'}


### Fit a SVM classification model using Random Search for Linear kernel

In [7]:
score_measure = "f1"
kfolds = 5

param_grid = {
     'C': [0.0001, 0.001, 0.1, 1, 10, 100, 1000], 
    'kernel': ['linear']
}

rand_linear_SVC = SVC()
rand_search = RandomizedSearchCV(estimator = rand_linear_SVC, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 7 is smaller than n_iter=500. Running 7 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 7 candidates, totalling 35 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9724053296218781
... with parameters: {'kernel': 'linear', 'C': 1}


### Fit a SVM classification model using Grid Search for Linear kernel

In [8]:
score_measure = "f1"
kfolds = 10

param_grid = {
     'C': [0.0001, 0.001, 0.1, 1, 10, 100, 1000], 
    'kernel': ['linear']
}

Grid_Linear_SVC = SVC()
grid_search = GridSearchCV(estimator = Grid_Linear_SVC, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 10 folds for each of 7 candidates, totalling 70 fits
The best f1 score is 0.9776157863464366
... with parameters: {'C': 10, 'kernel': 'linear'}


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


### Fit a SVM classification model using Random Search for Poly kernel

In [9]:
score_measure = "f1"
kfolds = 10

param_grid = {
     'C': [0.1,1, 10, 100], 
    'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}

Random_Poly_SVC = SVC()
rand_search = RandomizedSearchCV(estimator = Random_Poly_SVC, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 16 is smaller than n_iter=500. Running 16 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 10 folds for each of 16 candidates, totalling 160 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9736799574694313
... with parameters: {'kernel': 'poly', 'gamma': 1, 'C': 0.1}


### Fit a SVM classification model using Grid Search for Poly kernel

In [10]:
score_measure = "f1"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
    'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}


Grid_Poly_SVC = SVC()
grid_search = GridSearchCV(estimator = Grid_Poly_SVC, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 16 candidates, totalling 80 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9724053296218781
... with parameters: {'C': 0.1, 'gamma': 1, 'kernel': 'poly'}


### Fit a SVM classification model using Random Search for RBF kernel

In [12]:
score_measure = "f1"
kfolds = 10

param_grid = {
     'C': [0.1,1, 10], 
    'gamma': [1,0.1,0.011],
    'kernel': ['rbf']
}

Random_rbf_SVC = SVC()
rand_search = RandomizedSearchCV(estimator = Random_rbf_SVC, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 9 is smaller than n_iter=500. Running 9 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 10 folds for each of 9 candidates, totalling 90 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9757976045282547
... with parameters: {'kernel': 'rbf', 'gamma': 0.1, 'C': 10}


### Fit a SVM classification model using Grid Search for rbf kernel

In [13]:
score_measure = "f1"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,100),  
    'min_samples_leaf': np.arange(1,100),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 100), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 250, in fit
    raise ValueError(
ValueError: min_samples_split must be an integer greater than 1 or 

The best f1 score is 0.9872873200854851
... with parameters: {'min_samples_split': 6, 'min_samples_leaf': 2, 'min_impurity_decrease': 0.004600000000000001, 'max_leaf_nodes': 78, 'max_depth': 14, 'criterion': 'entropy'}


## Initial random search based on the parameters set for the Random search for the Decision Tree Classifier

In [14]:
score_measure = "f1"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,100),  
    'min_samples_leaf': np.arange(1,100),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 100), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
20 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
20 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 250, in fit
    raise ValueError(
ValueError: min_samples_split must be an integer greater than 1 or 

The best f1 score is 0.9891221824708063
... with parameters: {'min_samples_split': 13, 'min_samples_leaf': 5, 'min_impurity_decrease': 0.0091, 'max_leaf_nodes': 11, 'max_depth': 33, 'criterion': 'entropy'}


## Exhaustive grid search based on the parameters set for the Grid search for the Decision Tree Classifier

In [15]:
score_measure = "f1"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(10,15),  
    'min_samples_leaf': np.arange(3,7),
    'min_impurity_decrease': np.arange(0.0089, 0.0095, 0.0001),
    'max_leaf_nodes': np.arange(8,13), 
    'max_depth': np.arange(30,35), 
    'criterion': ['entropy'],
}

dtree_grid = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = dtree_grid, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 3000 candidates, totalling 15000 fits
The best f1 score is 0.9891221824708063
... with parameters: {'criterion': 'entropy', 'max_depth': 30, 'max_leaf_nodes': 8, 'min_impurity_decrease': 0.0089, 'min_samples_leaf': 3, 'min_samples_split': 10}


## Neural Network using MLP Classifier

### With Random Search

In [16]:
%%time

score_measure = "f1"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (50,), (70,),(50,30), (40,20), (60,40, 20), (70,50,40)],
    'activation': ['logistic', 'tanh', 'relu'],
    'solver': ['adam', 'sgd'],
    'alpha': [0, .2, .5, .7, 1],
    'learning_rate': ['constant', 'invscaling', 'adaptive'],
    'learning_rate_init': [0.001, 0.01, 0.1, 0.2, 0.5],
    'max_iter': [5000]
}

ann = MLPClassifier()
rand_search = RandomizedSearchCV(estimator = ann, param_distributions=param_grid, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

Fitting 5 folds for each of 100 candidates, totalling 500 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9759743531909016
... with parameters: {'solver': 'adam', 'max_iter': 5000, 'learning_rate_init': 0.1, 'learning_rate': 'constant', 'hidden_layer_sizes': (40, 20), 'alpha': 1, 'activation': 'relu'}
CPU times: total: 5.73 s
Wall time: 35min 27s


### With Grid Search

In [17]:
%%time

score_measure = "f1"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (30,), (50,), (70,), (90,)],
    'activation': ['tanh', 'relu'],
    'solver': ['adam'],
    'alpha': [.5, .7, 1],
    'learning_rate': ['constant', 'adaptive', 'invscaling'],
    'learning_rate_init': [0.001, 0.005, 0.01, 0.15],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = GridSearchCV(estimator = ann, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 288 candidates, totalling 1440 fits


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best f1 score is 0.9777758544419441
... with parameters: {'activation': 'tanh', 'alpha': 1, 'hidden_layer_sizes': (70,), 'learning_rate': 'adaptive', 'learning_rate_init': 0.01, 'max_iter': 5000, 'solver': 'adam'}
CPU times: total: 20.2 s
Wall time: 49min 39s


In [41]:
import os
import tensorflow as tf
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
from __future__ import print_function

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score

import tensorflow as tf
from tensorflow import keras

In [42]:
%%time

# create model stucture
model = keras.models.Sequential()
model.add(keras.layers.Input(1472))
model.add(keras.layers.Dense(50, activation='relu', kernel_initializer='glorot_normal'))
model.add(keras.layers.Dense(50, activation='relu', kernel_initializer='glorot_normal'))
model.add(keras.layers.Dense(50, activation='relu', kernel_initializer='glorot_normal'))
model.add(keras.layers.Dense(10, activation='softmax', kernel_initializer='glorot_normal')) # final layer, 10 categories


# compile
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

CPU times: total: 234 ms
Wall time: 6.58 s


In [43]:
%%time

# fit the model
history = model.fit(X_train, y_train, 
                    validation_data=(X_test, y_test), 
                    epochs=20, batch_size=100)

Epoch 1/20
6/6 [==============================] - 30s 4s/step - loss: 2.2968 - accuracy: 0.1423 - val_loss: 2.2730 - val_accuracy: 0.0773
Epoch 2/20
6/6 [==============================] - 5s 894ms/step - loss: 2.2246 - accuracy: 0.5255 - val_loss: 2.2014 - val_accuracy: 0.1078
Epoch 3/20
6/6 [==============================] - 3s 548ms/step - loss: 2.1230 - accuracy: 0.5620 - val_loss: 2.0861 - val_accuracy: 0.0957
Epoch 4/20
6/6 [==============================] - 2s 403ms/step - loss: 1.9480 - accuracy: 0.5584 - val_loss: 1.8809 - val_accuracy: 0.1035
Epoch 5/20
6/6 [==============================] - 2s 387ms/step - loss: 1.6617 - accuracy: 0.5620 - val_loss: 1.5507 - val_accuracy: 0.1367
Epoch 6/20
6/6 [==============================] - 3s 548ms/step - loss: 1.2616 - accuracy: 0.5985 - val_loss: 1.1239 - val_accuracy: 0.2508
Epoch 7/20
6/6 [==============================] - 2s 376ms/step - loss: 0.8721 - accuracy: 0.7263 - val_loss: 0.7781 - val_accuracy: 0.5456
Epoch 8/20
6/6 [======

In [45]:
# evaluate the model

scores = model.evaluate(X_test, y_test, verbose=0)
scores
# In results, first is loss, second is precision

[0.23477661609649658, 0.9517529010772705]

In [48]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.23
accuracy: 95.18%


In [49]:
model = keras.models.Sequential()

model.add(keras.layers.Input(shape=1472))
model.add(keras.layers.Dense(100, activation='relu', kernel_initializer='glorot_normal'))
model.add(keras.layers.Dense(100, activation='relu', kernel_initializer='glorot_normal'))
model.add(keras.layers.Dense(100, activation='relu', kernel_initializer='glorot_normal'))
model.add(keras.layers.Dense(10, activation='softmax', kernel_initializer='glorot_normal'))

In [52]:
import keras.backend as K

def precision(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
    precision = true_positives / (predicted_positives + K.epsilon())
    return precision

In [55]:
adam = keras.optimizers.Adam(learning_rate=0.01)
model.compile(loss='sparse_categorical_crossentropy', optimizer=adam, metrics=[precision])

history = model.fit(X_train, y_train, 
                    validation_data=(X_test, y_test), 
                    epochs=20, batch_size=100)

Epoch 1/20
6/6 [==============================] - 39s 4s/step - loss: 1.5455 - precision: 0.2090 - val_loss: 0.8405 - val_precision: 0.0026
Epoch 2/20
6/6 [==============================] - 4s 783ms/step - loss: 0.5953 - precision: 0.5018 - val_loss: 0.2480 - val_precision: 0.0026
Epoch 3/20
6/6 [==============================] - 3s 552ms/step - loss: 0.2056 - precision: 0.4982 - val_loss: 0.2083 - val_precision: 0.0026
Epoch 4/20
6/6 [==============================] - 3s 551ms/step - loss: 0.0369 - precision: 0.4982 - val_loss: 0.1261 - val_precision: 0.0026
Epoch 5/20
6/6 [==============================] - 2s 418ms/step - loss: 0.0225 - precision: 0.4928 - val_loss: 0.2424 - val_precision: 0.0026
Epoch 6/20
6/6 [==============================] - 3s 547ms/step - loss: 0.0119 - precision: 0.4910 - val_loss: 0.1753 - val_precision: 0.0026
Epoch 7/20
6/6 [==============================] - 2s 389ms/step - loss: 0.0143 - precision: 0.5072 - val_loss: 0.4481 - val_precision: 0.0026
Epoch 8/

In [56]:
scores = model.evaluate(X_test, y_test, verbose=0)
scores

[0.8151329159736633, 0.0026070252060890198]

In [57]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.82
precision: 0.26%


In [58]:
#Random search

In [68]:
%%time

# If you don't have the following installed, from command line '!pip install scikeras'
from scikeras.wrappers import KerasClassifier
from keras.initializers import GlorotNormal

score_measure = "precision"
kfolds = 5

def build_clf(hidden_layer_sizes, dropout):
    ann = tf.keras.models.Sequential()
    ann.add(keras.layers.Input(shape=1472)),
    for hidden_layer_size in hidden_layer_sizes:
        model.add(keras.layers.Dense(hidden_layer_size, kernel_initializer= tf.keras.initializers.GlorotUniform(), 
                                     bias_initializer=keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None), activation="relu"))
        model.add(keras.layers.Dropout(dropout))
    ann.add(tf.keras.layers.Dense(10, activation='softmax',  kernel_initializer='glorot_normal' ))
    ann.compile(loss = 'sparse_categorical_crossentropy', metrics = ['precision'])
    return ann

CPU times: total: 0 ns
Wall time: 0 ns


In [69]:
from scikeras.wrappers import KerasClassifier

keras_clf = KerasClassifier(
    model=build_clf,
    hidden_layer_sizes=1472,
    dropout = 0.0
)

In [70]:
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.model_selection import RandomizedSearchCV

params = {
    'optimizer__learning_rate': [0.0005, 0.001, 0.005],
    'model__hidden_layer_sizes': [(70,),(90, ), (100,), (100, 90)],
    'model__dropout': [0, 0.1],
    'batch_size':[20, 60, 100],
    'epochs':[10, 50, 100],
    'optimizer':["adam",'sgd']
}
keras_clf.get_params().keys()


dict_keys(['model', 'build_fn', 'warm_start', 'random_state', 'optimizer', 'loss', 'metrics', 'batch_size', 'validation_batch_size', 'verbose', 'callbacks', 'validation_split', 'shuffle', 'run_eagerly', 'epochs', 'hidden_layer_sizes', 'dropout', 'class_weight'])

In [72]:
import keras.backend as K



rnd_search_cv = RandomizedSearchCV(estimator=keras_clf, param_distributions=params, scoring='precision', n_iter=50, cv=5)

import sys
sys.setrecursionlimit(10000) # note: the default is 3000 (python 3.9)

earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=0, mode='auto')
callback = [earlystop]

_ = rnd_search_cv.fit(X_train, y_train, callbacks=callback, verbose=0)


C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
250 fits failed out of a total of 250.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
250 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\scikeras\wrappers.py", line 1494, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "C:\Users\Srinidhi\anaconda3\lib\site-packages\scikeras\wrappers.py", line 762, in fit
    self._fit(
  File "C:\Users\Srinidhi\anacon

ValueError: in user code:

    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\training.py", line 998, in train_step
        return self.compute_metrics(x, y, y_pred, sample_weight)
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\training.py", line 1092, in compute_metrics
        self.compiled_metrics.update_state(y, y_pred, sample_weight)
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\compile_utils.py", line 577, in update_state
        self.build(y_pred, y_true)
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\compile_utils.py", line 483, in build
        self._metrics = tf.__internal__.nest.map_structure_up_to(
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\compile_utils.py", line 631, in _get_metric_objects
        return [self._get_metric_object(m, y_t, y_p) for m in metrics]
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\compile_utils.py", line 631, in <listcomp>
        return [self._get_metric_object(m, y_t, y_p) for m in metrics]
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\engine\compile_utils.py", line 650, in _get_metric_object
        metric_obj = metrics_mod.get(metric)
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\metrics\__init__.py", line 181, in get
        return deserialize(str(identifier))
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\metrics\__init__.py", line 136, in deserialize
        return deserialize_keras_object(
    File "C:\Users\Srinidhi\anaconda3\lib\site-packages\keras\utils\generic_utils.py", line 769, in deserialize_keras_object
        raise ValueError(

    ValueError: Unknown metric function: precision. Please ensure this object is passed to the `custom_objects` argument. See https://www.tensorflow.org/guide/keras/save_and_serialize#registering_the_custom_object for details.


In [ ]:
sdbvcvcrnd_search_cv.best_params_sdbvcvcrnd_search_cv.best_params_sdbvcvcrnd_search_cv.best_params_sdbvcvcccscssssssssaaaasaaaaaxxsacasxaasxxcv ccccxaccc vvvvvvvvvvvvvvvvvvvvvdbbbbbbccv cxzx cc cczx  z cv  xxxxxxxxxxxxxxxxxxcccCZCrnd_search_cv.best_params_

In [ ]:
best_net = rnd_search_cv.best_estimator_
print(rnd_search_cv.best_params_)

In [ ]:
%%time
y_pred = best_net.predict(X_test)
print(classification_report(y_test, y_pred))

## Inference:

1. The f1 score for Random Search on Logistic Regression is 0.9762289562289563

2. The f1 score for grid search on the logistic regression is 0.9742068308729206

3. The f1 score for Random Search on Linear SVM is 0.9724053296218781

4. The f1 score for Grid Search on Linear SVM is 0.9776157863464366

5. The f1 score for Random Search on Poly SVM is 0.9736799574694313

6. The f1 score for Grid Search on Poly SVM is 0.9724053296218781

7. The f1 score for Random Search on RBF SVM is 0.9757976045282547

8. The f1 score for Grid Search on RBF SVM is 0.9872873200854851

9. The f1 score for Random Search on Decision tree classifier is 0.9891221824708063

10. The f1 score for exhaustive Grid Search on Decision tree classifier is 0.9891221824708063

11. The f1 score for Random Search on NN MLP Classifier is 0.9759743531909016

12. The f1 score for Grid Search on on NN MLP Classifier is 0.9777758544419441

When I try to include the Neural Networks using the MLP classifier into the model performing notebook, I tend to see that the NN is giving  the best f1 score of 97.59 for the Random Search training and 97.77 for the Grid Search. However when we compare the Grid Search scores of the MLP Classifier NN with that of the Grid search of the Decision Tree classifier or the RBF SVM classifier, it was comparitively low in the F1 score. This doesn't necessarily mean that the NN is not the better model but its not giving the best results compared to the decision tree or the SVM for this particular problem while fairly giving out the better end of the results. While the Neural Networks is a powerful model that works best for the complex problems,  I have tried my best to make the dataset as easier to analyse as possible by eliminating the complex relations in the dataset such as making the continuous variables which would have worked best with the MLP regressor and making them classified and also elimating the variables that could very well contribute to the complexity in the dataset; which might have worked for the NN to learn from this data; it could be possible that the Decision tree was able to predict the best results as it produces a tree like structure that can easily be analyzed. In conclusion, these results apply only for this dataset and doesn't necessarily be applied for the raw dataset where the more of the regression type of models are used. But based on this dataset results, I can certainly say that by including the MLP classifier, I did not see any change of the results from the previous iteration and the Decision Tree is still the best performing model compared to the rest of the models.